src/animation/event.rs

```diff
@@ -0,0 +1,163 @@
+//! This module animition event.
+//!
+
+use bevy::{prelude::*, reflect::GetTypeRegistration, utils::HashMap};
+
+use crate::asset::AnimationClip2D;
+
+use super::{
+    animation_spritesheet::animation_player_spritesheet, AnimationPlayer2D,
+    AnimationPlayer2DSystemSet,
+};
+
+/// AnimationEvents are triggered by the animation system if registered as such with the App.
+pub trait AnimationEvent: Event + GetTypeRegistration + FromReflect + Clone {
+    /// Implement this to be able to set the entity for a targeted event.
+    /// Default implementation is a No-Op.
+    fn set_target(&mut self, target: EventTarget) {
+        let _ = target;
+        /* Default implementation is empty for non-targeted events */
+    }
+}
+
+/// Wrapper around entity to be used for EventTargets
+#[derive(Debug, Clone, Copy, Deref, Reflect)] //TODO: register type
+pub struct EventTarget(pub Entity);
+
+impl Default for EventTarget {
+    fn default() -> Self {
+        Self(Entity::PLACEHOLDER)
+    }
+}
+
+#[derive(Debug, Resource)]
+struct AnimationEventCache<T>(HashMap<AssetId<AnimationClip2D>, HashMap<usize, Vec<T>>>);
+
+impl<T> Default for AnimationEventCache<T> {
+    fn default() -> Self {
+        Self(Default::default())
+    }
+}
+
+// This updates a cache resource for each Event added to the app
+// That way when processing animation for event sending, we already have a vector of T instead of Box<dyn Reflect>, so we only iterate through the events that are actually relecant (can be from_reflected to T)
+fn update_animation_event_cache<T: FromReflect>(
+    mut cache: ResMut<AnimationEventCache<T>>,
+    mut asset_events: EventReader<AssetEvent<AnimationClip2D>>,
+    animation_clips: Res<Assets<AnimationClip2D>>,
+) {
+    for asset_event in asset_events.read() {
+        match asset_event {
+            AssetEvent::Added { id }
+            | AssetEvent::Modified { id }
+            | AssetEvent::LoadedWithDependencies { id } => {
+                let clip = animation_clips.get(*id).unwrap();
```

Comment:
Weird, I am like at least 80% sure I tested this, which is a lot for me :D.
Looking at it, maybe Added is not ensured to be loaded? I may be able to take look tomorrow.

Great you are starting to use it :)

Reply:
Hahaha, yeah I get that. It's also getting late for so I am not too sure why this is happening or if I have perhaps done an oopsie.

Yeah the events are a pretty neat solution I think, although testing the events without the debouncing is gonna be a little difficult, so I probably want be able to properly implement them into my project until then.

Reply:
Oh yeah also, I had to merge main into this branch because I am using the new duration stuff in my project, this is the branch I am referencing in my game https://github.com/PraxTube/bevy_trickfilm/tree/testy.

Reply:
Alright so this fixes the issue, I am guessing there is some kind of one frame delay or something by loading the assets or whatever, and it doesn't happen in the examples here because they are super tiny, but my project takes longer and that's why it is invalid for a few frames (is my guess). https://github.com/PraxTube/bevy_trickfilm/commit/a0b8028b0586d61bc1301f860952d23019a5b319

Still not done though, I can't quite the parsing of the events to work.

```
2024-10-24T05:56:37.778738Z ERROR bevy_asset::server: Failed to load asset 'dude/dude.trickfilm.ron' with asset loader 'bevy_trickfilm::asset::asset_loader::Animation2DLoader': Could not parse RON: 199:31: No registration found for `events::SpawnHitboxEvent`
```

I have created `src/events.rs` though doesn't work, I also tried in different locations but nothing seemed to work. I don't quite understand this parsing it seems like, I thought it's the path to the event, but then I changed the example `events` to `testy` and also changed `events::SampleEvent` to `testy::SampleEvent` and got the same error.

Reply:
The path needs to start with the crate name. For examples I think it take the name not from the source file, but the toml.
Do you have a link to your project so I can check?

Reply:
https://github.com/PraxTube/rantoni/tree/implement-new-trickfilm-events

`dude.trickfilm.ron` has the event animation data and `events.rs` defines the even struct.

Reply:
Woah, your game is awesome.
You just forgot to add your AssetEventsPlugin.

Reply:
Haha thanks, though it's still very much a prototype.

Omg yeah... Okay yeah so I was just missing the name of the project, should definitely add some comments or notes regarding this as it wasn't obvious to me at all.

Alright, awesome. It works really nice, it also reduces code on my side quite a bit and allows me to save animation data all in one place which is great. I didn't test the observers as I don't use them (yet).